core/shared/src/main/scala/zio/prelude/Semiring.scala

```diff
@@ -0,0 +1,398 @@
+package zio.prelude
+
+import scala.annotation.tailrec
+
+/**
+ * `Semiring` is a data type that represents some notion of "events" that can
+ * take place in parallel or in sequence. For example, a semiring
+ * parameterized on some error type could be used to model the potentially
+ * multiple ways that an application can fail. On the other hand, a semiring
+ * parameterized on some request type could be used to model a collection of
+ * requests to external data sources, some of which could be executed in
+ * parallel and some of which must be executed sequentially.
+ */
+sealed trait Semiring[+Z <: Unit, +A] { self =>
+
+  /**
+   * Combines this collection of events with that collection of events to
+   * return a new collection of events that represents this collection of
+   * events in parallel with that collection of events.
+   */
+  final def &&[Z1 >: Z <: Unit, A1 >: A](that: Semiring[Z1, A1]): Semiring[Z1, A1] =
```

Comment:
We're losing information right now that combining an empty collection with a nonempty collection results in a nonempty collection. Potentially we could could use overloading or type level programming to recover this.